File: core/src/main/com/thymeflow/rdf/sail/SailInterceptor.java
package com.thymeflow.rdf.sail;

import org.eclipse.rdf4j.model.IRI;
import org.eclipse.rdf4j.model.Resource;
import org.eclipse.rdf4j.model.Value;

/**
 * @author Thomas Pellissier Tanon
 */
public interface SailInterceptor {
    /**
     * Method called when an insertion is tried on the sail
     *
     * @return boolean if the insertion should be done
     */
    default boolean onSparqlAddStatement(Resource subject, IRI predicate, Value object, Resource... contexts) {
        return true;
    }

    /**
     * Method called when an deletion is tried on the sail
     *
     * @return boolean if the deletion should be done
     */
    default boolean onSparqlRemoveStatement(Resource subject, IRI predicate, Value object, Resource... contexts) {
        return true;
    }
}

File: core/src/main/com/thymeflow/rdf/sail/SailInterceptor.scala
package com.thymeflow.rdf.sail

import com.thymeflow.rdf.model.StatementSetDiff

/**
  * @author Thomas Pellissier Tanon
  */
trait SailInterceptor {
  /**
    * Method called when an SPARQL UPDATE is tried on the sail
    */
  def onSparqlUpdateDiff(statementSetDiff: StatementSetDiff)
}